File: common/rfb/CSecurityDH.cxx
/* 
 * Copyright (C) 2022 Dinglan Peng
 *    
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#ifndef HAVE_NETTLE
#error "This header should not be compiled without HAVE_NETTLE defined"
#endif

#include <stdlib.h>
#ifndef WIN32
#include <unistd.h>
#endif
#include <assert.h>

#include <nettle/aes.h>
#include <nettle/md5.h>
#include <nettle/bignum.h>
#include <rfb/CSecurityDH.h>
#include <rfb/CConnection.h>
#include <rdr/InStream.h>
#include <rdr/OutStream.h>
#include <rdr/RandomStream.h>
#include <rfb/Exception.h>
#include <os/os.h>

using namespace rfb;

const int MinKeyLength = 128;
const int MaxKeyLength = 1024;

CSecurityDH::CSecurityDH(CConnection* cc_)
  : CSecurity(cc_), keyLength(0)
{
  mpz_init(g);
  mpz_init(p);
  mpz_init(A);
  mpz_init(b);
  mpz_init(B);
  mpz_init(k);
}

CSecurityDH::~CSecurityDH()
{
  mpz_clear(g);
  mpz_clear(p);
  mpz_clear(A);
  mpz_clear(b);
  mpz_clear(B);
  mpz_clear(k);
}

bool CSecurityDH::processMsg()
{
  if (readKey()) {
    writeCredentials();
    return true;
  }
  return false;
}

bool CSecurityDH::readKey()
{
  rdr::InStream* is = cc->getInStream();
  if (!is->hasData(4))
    return false;
  is->setRestorePoint();
  uint16_t gen = is->readU16();
  keyLength = is->readU16();
  if (keyLength < MinKeyLength)
    throw Exception("DH key is too short");
  if (keyLength > MaxKeyLength)
    throw Exception("DH key is too long");
  if (!is->hasDataOrRestore(keyLength * 2))
    return false;
  is->clearRestorePoint();
  mpz_set_ui(g, gen);
  std::vector<uint8_t> pBytes(keyLength);
  std::vector<uint8_t> ABytes(keyLength);
  is->readBytes(pBytes.data(), pBytes.size());
  is->readBytes(ABytes.data(), ABytes.size());
  nettle_mpz_set_str_256_u(p, pBytes.size(), pBytes.data());
  nettle_mpz_set_str_256_u(A, ABytes.size(), ABytes.data());
  return true;
}

void CSecurityDH::writeCredentials()
{
  std::string username;
  std::string password;
  rdr::RandomStream rs;

  (CSecurity::upg)->getUserPasswd(isSecure(), &username, &password);

  std::vector<uint8_t> bBytes(keyLength);
  if (!rs.hasData(keyLength))
    throw Exception("failed to generate DH private key");
  rs.readBytes(bBytes.data(), bBytes.size());
  nettle_mpz_set_str_256_u(b, bBytes.size(), bBytes.data());
  mpz_powm(k, A, b, p);
  mpz_powm(B, g, b, p);

  std::vector<uint8_t> sharedSecret(keyLength);
  std::vector<uint8_t> BBytes(keyLength);
  nettle_mpz_get_str_256(sharedSecret.size(), sharedSecret.data(), k);
  nettle_mpz_get_str_256(BBytes.size(), BBytes.data(), B);
  uint8_t key[16];
  struct md5_ctx md5Ctx;
  md5_init(&md5Ctx);
  md5_update(&md5Ctx, sharedSecret.size(), sharedSecret.data());
  md5_digest(&md5Ctx, 16, key);
  struct aes128_ctx aesCtx;
  aes128_set_encrypt_key(&aesCtx, key);

  uint8_t buf[128];
  if (!rs.hasData(128))
    throw Exception("failed to generate random padding");
  rs.readBytes(buf, 128);
  if (username.size() >= 64)
    throw Exception("username is too long");
  memcpy(buf, username.c_str(), username.size() + 1);
  if (password.size() >= 64)
    throw Exception("password is too long");
  memcpy(buf + 64, password.c_str(), password.size() + 1);
  aes128_encrypt(&aesCtx, 128, buf, buf);

  rdr::OutStream* os = cc->getOutStream();
  os->writeBytes(buf, 128);
  os->writeBytes(BBytes.data(), BBytes.size());
  os->flush();
}

File: java/com/tigervnc/rfb/CSecurityDH.java
/* Copyright (C) 2022 Dinglan Peng
 * Copyright (C) 2026 Brian P. Hinz
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */

package com.tigervnc.rfb;

import java.io.UnsupportedEncodingException;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;

import com.tigervnc.rdr.*;

public class CSecurityDH extends CSecurity {

  private static final int MinKeyLength = 128;
  private static final int MaxKeyLength = 1024;

  private static byte[] bigIntToBytes(BigInteger n, int bytes) {
    byte[] arr = n.toByteArray();
    int len = arr.length < bytes ? arr.length : bytes;
    byte[] res = new byte[bytes];
    System.arraycopy(arr, arr.length - len, res, bytes - len, len);
    return res;
  }

  public boolean processMsg(CConnection cc) {
    readKey(cc);
    writeCredentials(cc);
    return true;
  }

  private void readKey(CConnection cc) {
    InStream is = cc.getInStream();

    int gen = is.readU16();
    keyLength = is.readU16();
    if (keyLength < MinKeyLength)
      throw new AuthFailureException("Received Diffie-Hellman key is too short");
    if (keyLength > MaxKeyLength)
      throw new AuthFailureException("Received Diffie-Hellman key is too long");

    byte[] pBytes = new byte[keyLength];
    byte[] aBytes = new byte[keyLength];
    is.readBytes(ByteBuffer.wrap(pBytes), keyLength);
    is.readBytes(ByteBuffer.wrap(aBytes), keyLength);

    g = BigInteger.valueOf(gen);
    p = new BigInteger(1, pBytes);
    A = new BigInteger(1, aBytes);

    // A modulus of 0 or 1 is degenerate: BigInteger.modPow() would either
    // throw ArithmeticException (0, or negative) or silently derive an
    // always-zero, trivially-predictable shared secret (1), so reject
    // both explicitly rather than let either happen deep in the key
    // exchange (see GHSA-r4vv-jmq7-c5ph, filed against the C++ viewer's
    // equivalent unchecked-modulus code).
    if (p.compareTo(BigInteger.ONE) <= 0)
      throw new AuthFailureException("Received invalid Diffie-Hellman modulus");
  }

  private void writeCredentials(CConnection cc) {
    StringBuffer username = new StringBuffer();
    StringBuffer password = new StringBuffer();
    CSecurity.upg.getUserPasswd(isSecure(), username, password);

    SecureRandom rand = new SecureRandom();

    byte[] bBytes = new byte[keyLength];
    rand.nextBytes(bBytes);
    BigInteger b = new BigInteger(1, bBytes);

    BigInteger k = A.modPow(b, p);
    BigInteger B = g.modPow(b, p);

    byte[] sharedSecret = bigIntToBytes(k, keyLength);
    byte[] BBytes = bigIntToBytes(B, keyLength);

    byte[] key;
    try {
      MessageDigest md5 = MessageDigest.getInstance("MD5");
      key = md5.digest(sharedSecret);
    } catch (NoSuchAlgorithmException e) {
      throw new AuthFailureException("MD5 algorithm is not supported");
    }

    // The 128-byte credentials buffer is random-padded, matching the
    // server's expectation that unused space not be predictable/zeroed.
    byte[] buf = new byte[128];
    rand.nextBytes(buf);

    byte[] usernameBytes;
    byte[] passwordBytes;
    try {
      usernameBytes = username.toString().getBytes("UTF-8");
      passwordBytes = password.toString().getBytes("UTF-8");
    } catch (UnsupportedEncodingException e) {
      throw new AuthFailureException("UTF-8 is not supported");
    }
    if (usernameBytes.length >= 64)
      throw new AuthFailureException("Username is too long");
    if (passwordBytes.length >= 64)
      throw new AuthFailureException("Password is too long");
    System.arraycopy(usernameBytes, 0, buf, 0, usernameBytes.length);
    buf[usernameBytes.length] = 0;
    System.arraycopy(passwordBytes, 0, buf, 64, passwordBytes.length);
    buf[64 + passwordBytes.length] = 0;

    byte[] encrypted;
    try {
      Cipher cipher = Cipher.getInstance("AES/ECB/NoPadding");
      cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, "AES"));
      encrypted = cipher.doFinal(buf);
    } catch (java.lang.Exception e) {
      throw new AuthFailureException("Failed to encrypt DH credentials: "+e.getMessage());
    }

    OutStream os = cc.getOutStream();
    os.writeBytes(encrypted, 0, encrypted.length);
    os.writeBytes(BBytes, 0, BBytes.length);
    os.flush();
  }

  public int getType() { return Security.secTypeDH; }
  public String description() { return "Diffie-Hellman"; }

  private int keyLength;
  private BigInteger g, p, A;
}
